drivers/comparator/comparator_nrf_comp.c

```diff
@@ -0,0 +1,771 @@
+	}
+
+	nrf->interrupt_priority = 0;
+	return 0;
+}
+
+static int shim_nrf_comp_diff_config_to_nrf(const struct comp_nrf_comp_diff_config *shim,
+					    nrfx_comp_config_t *nrf)
+{
+	nrf->reference = NRF_COMP_REF_AREF;
+
+	if (shim_nrf_comp_extrefsel_to_nrf(shim->extrefsel, &nrf->ext_ref)) {
+		return -EINVAL;
+	}
+
+	nrf->main_mode = NRF_COMP_MAIN_MODE_DIFF;
+	nrf->threshold.th_down = 0;
+	nrf->threshold.th_up = 0;
+
+	if (shim_nrf_comp_sp_mode_to_nrf(shim->sp_mode, &nrf->speed_mode)) {
+		return -EINVAL;
+	}
+
+#if NRF_COMP_HAS_HYST
+	if (shim->enable_hyst) {
+		nrf->hyst = NRF_COMP_HYST_ENABLED;
+	} else {
+		nrf->hyst = NRF_COMP_HYST_DISABLED;
+	}
+#else
+	if (shim->enable_hyst) {
+		return -EINVAL;
+	}
+#endif
+
+#if NRF_COMP_HAS_ISOURCE
+	if (shim_nrf_comp_isource_to_nrf(shim->isource, &nrf->isource)) {
+		return -EINVAL;
+	}
+#else
+	if (shim->isource != COMP_NRF_COMP_ISOURCE_DISABLED) {
+		return -EINVAL;
+	}
+#endif
+
+	if (shim_nrf_comp_psel_to_nrf(shim->psel, &nrf->input)) {
+		return -EINVAL;
+	}
+
+	nrf->interrupt_priority = 0;
+	return 0;
+}
+
+static int shim_nrf_comp_get_output(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return nrfx_comp_sample();
+}
+
+static int shim_nrf_comp_set_trigger(const struct device *dev,
+				     enum comparator_trigger trigger)
+{
+	shim_nrf_comp_stop();
+
+	switch (trigger) {
+	case COMPARATOR_TRIGGER_NONE:
+		shim_nrf_comp_data0.event_mask = 0;
+		break;
+
+	case COMPARATOR_TRIGGER_RISING_EDGE:
+		shim_nrf_comp_data0.event_mask = NRF_COMP_INT_UP_MASK;
+		break;
+
+	case COMPARATOR_TRIGGER_FALLING_EDGE:
+		shim_nrf_comp_data0.event_mask = NRF_COMP_INT_DOWN_MASK;
+		break;
+
+	case COMPARATOR_TRIGGER_BOTH_EDGES:
+		shim_nrf_comp_data0.event_mask = NRF_COMP_INT_CROSS_MASK;
+		break;
+	}
+
+	if (shim_nrf_comp_is_resumed()) {
+		shim_nrf_comp_start();
+	}
+
+	return 0;
+}
+
+static int shim_nrf_comp_set_trigger_callback(const struct device *dev,
+					      comparator_callback_t callback,
+					      void *user_data)
+{
+	shim_nrf_comp_stop();
+
+	shim_nrf_comp_data0.callback = callback;
+	shim_nrf_comp_data0.user_data = user_data;
+
+	if (callback != NULL && atomic_test_and_clear_bit(&shim_nrf_comp_data0.triggered, 0)) {
+		callback(dev, user_data);
+	}
+
+	if (shim_nrf_comp_is_resumed()) {
+		shim_nrf_comp_start();
+	}
+
+	return 0;
+}
+
+static int shim_nrf_comp_trigger_is_pending(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return atomic_test_and_clear_bit(&shim_nrf_comp_data0.triggered, 0);
+}
+
+static const struct comparator_driver_api shim_nrf_comp_api = {
+	.get_output = shim_nrf_comp_get_output,
+	.set_trigger = shim_nrf_comp_set_trigger,
+	.set_trigger_callback = shim_nrf_comp_set_trigger_callback,
+	.trigger_is_pending = shim_nrf_comp_trigger_is_pending,
+};
+
+static int shim_nrf_comp_reconfigure(const nrfx_comp_config_t *nrf)
+{
+	shim_nrf_comp_stop();
+
+	(void)nrfx_comp_reconfigure(nrf);
+
+	if (shim_nrf_comp_is_resumed()) {
+		shim_nrf_comp_start();
+	}
```

Comment:
If I understand it correctly, reconfiguration is allowed while the comparator is suspended. The problem here is that an assertion in `nrfx_comp_stop()` will fail if the comparator is already stopped. I guess stopping should also be conditional.

Reply:
it is :) `shim_nrf_comp_stop();` and `shim_nrf_comp_start();` check if the state is already met by tracking the `.started` member:
```
static void shim_nrf_comp_start(void)
{
	if (shim_nrf_comp_data0.started) {
		return;
	}

	nrfx_comp_start(shim_nrf_comp_data0.event_mask, 0);
	shim_nrf_comp_data0.started = true;
}
```

Reply:
Right. I somehow missed that additional logic in `shim_nrf_comp_stop()`.